src/main/scala/com/fulcrumgenomics/bam/FindSwitchbackReads.scala

```diff
@@ -0,0 +1,398 @@
+  @arg(flag='O', doc="Maximum offset between end the two segments of the read on the reference. Set to 0 to disable read-based checks.") val maxOffset: Int = 35,
+  @arg(flag='g', doc="Maximum gap between R1 and R2 of tandem reads to call a template a switchback. Set to 0 to disable tandem-based checks.") val maxGap: Int = 500,
+  @arg(flag='l', doc="Minimum match length of the switched back segment.") val minLength: Int = 6,
+  @arg(flag='e', doc="Maximum error rate of switchback match to genome.") val maxErrorRate: Double = 0.1,
+  @arg(flag='d', doc="IF true, do NOT unmap reads from switchback templates.") val dontUnmap: Boolean = false
+) extends FgBioTool with LazyLogging {
+  import FindSwitchbackReads._
+  Io.assertReadable(input)
+  Io.assertReadable(ref)
+
+  override def execute(): Unit = {
+    logger.info("Loading reference sequences.")
+    val refs = ReferenceSequenceIterator(ref, stripComments=true).map(r => r.getName -> r).toMap
+
+    val progress = ProgressLogger(logger)
+    var total   = 0L
+    var aligned = 0L
+    val switchbackGaps    = new NumericCounter[Int]()
+    val switchbackLengths = new NumericCounter[Int]()
+    val switchbackOffsets = new NumericCounter[Int]()
+
+    val in = SamSource(input)
+    val iterator = Bams.templateIterator(in)
+
+    val outHeader = in.header.clone()
+    sortOrder match {
+      case Some(so) => so.applyTo(outHeader)
+      case None     =>
+        if (in.header.getSortOrder != SortOrder.queryname && in.header.getGroupOrder != GroupOrder.query) {
+          SamOrder.Queryname.applyTo(outHeader)
+        }
+    }
+    sortOrder.foreach(_.applyTo(outHeader))
+    val out = SamWriter(output, header=outHeader, sort=sortOrder)
+
+    logger.info("Examining reads for switchbacks.")
+    iterator.map { template =>
+      total += 1
+      if (!template.primaryReads.exists(_.mapped)) template else {
+        aligned += 1
+
+        val result: Option[SwitchHit] = findReadBasedSwitchback(refs, template, minLength, maxOffset, maxErrorRate)
+          .orElse(findTandemSwitchback(template, maxGap))
+
+        result match {
+          case None =>
+            Unit
+          case Some(hit: TandemBasedHit) =>
+            switchbackGaps.count(hit.gap)
+            val tagValue = s"${hit.code},${hit.gap}"
+            template.allReads.foreach(r => r(SwitchTag) = tagValue)
+          case Some(hit: ReadBasedHit) =>
+            switchbackLengths.count(hit.length)
+            switchbackOffsets.count(hit.offset)
+            val foundInReadNum = readNum(hit.read.getOrElse(unreachable("read should be populated!")))
+            template.allReads.foreach { r =>
+              val num = readNum(r)
+              r(SwitchTag) = s"${hit.code},${if (num == foundInReadNum) "read" else "mate"},${hit.offset},${hit.length}"
+            }
+        }
+
+        if (result.isDefined && !this.dontUnmap) template.unmapped else template
+      }
+    }
+    .foreach { template =>
```

Comment:
why does this need another `foreach` on the templates, why not just put it after the `if/else` that starts on 328?

Reply:
I just think it's clearer this way - since the output template varies based on the path above, it's easier/clear to `map().foreach()`.  The other option is to catch the return value of the `if` on line 328, and convert this into a single `foreach`.  I don't think that's any better.